map_model/src/pathfind/v2.rs

```diff
@@ -3,82 +3,181 @@
-                let turn = TurnID {
-                    parent: map.get_l(src).dst_i,
-                    src,
-                    dst,
-                };
-                if map.maybe_get_t(turn).is_some() {
-                    graph.add_edge(src, dst, turn);
-                }
-            }
+    cost: Duration,
+    // TODO Temporarily we'll keep plumbing these along for path_v2_to_v1 to work, but we'll
+    // probably just discover uber-turns lazily at the simulation layer instead.
+    uber_turns: Vec<UberTurnV2>,
+}
+
+impl PathV2 {
+    pub(crate) fn new(
+        steps: Vec<PathStepV2>,
+        req: PathRequest,
+        cost: Duration,
+        uber_turns: Vec<UberTurnV2>,
+    ) -> PathV2 {
+        // TODO Port validate_continuity and validate_restrictions?
+        PathV2 {
+            steps,
+            req,
+            cost,
+            uber_turns,
+        }
+    }
+
+    /// Vehicle implementations often just calculate the sequence of roads. Turn that into
+    /// PathStepV2 here.
+    pub(crate) fn from_roads(
+        mut roads: Vec<DirectedRoadID>,
+        req: PathRequest,
+        cost: Duration,
+        uber_turns: Vec<UberTurnV2>,
+        map: &Map,
+    ) -> PathV2 {
+        let mut steps = Vec::new();
+        for pair in roads.windows(2) {
+            steps.push(PathStepV2::Along(pair[0]));
+            steps.push(PathStepV2::Turn(MovementID {
+                from: pair[0],
+                to: pair[1],
+                parent: pair[0].dst_i(map),
+                crosswalk: false,
+            }));
         }
+        steps.push(PathStepV2::Along(roads.pop().unwrap()));
+        PathV2::new(steps, req, cost, uber_turns)
     }
 
-    match petgraph::algo::astar(
-        &graph,
-        req.start.lane(),
-        |l| l == req.end.lane(),
-        |(_, _, t)| {
-            // Normally opportunistic lane-changing adjusts the path live, but that doesn't work
-            // near uber-turns. So still use some of the penalties here.
-            let (lt, lc, slow_lane) = map.get_t(*t).penalty(map);
-            let mut extra_penalty = lt + lc;
-            if req.constraints == PathConstraints::Bike {
-                extra_penalty = slow_lane;
+    /// The original PathRequest used to produce this path.
+    pub fn get_req(&self) -> &PathRequest {
+        &self.req
+    }
+
+    /// All steps in this path.
+    pub fn get_steps(&self) -> &Vec<PathStepV2> {
+        &self.steps
+    }
+
+    /// The time needed to perform this path. This time is not a lower bound; physically following
+    /// the path might be faster. This time incorporates costs like using sub-optimal lanes or
+    /// taking difficult turns.
+    pub fn get_cost(&self) -> Duration {
+        self.cost
+    }
+
+    /// Transform a sequence of roads representing a path into the current lane-based path, by
+    /// picking particular lanes and turns to use.
+    pub fn to_v1(self, map: &Map) -> Result<Path> {
```

Comment:
Do you anticipate needing this only as a temporary compatibility shim for legacy code as you migrate everything to using road-based-paths, or do you expect these lane-based-paths to remain in the code base long term?

Reply:
I'm still not sure, but by the end of today, should have a much better idea. I'm hoping for temporary shim.